### PR TITLE
Generators now use same logic as Sequence

### PR DESCRIPTION
**Pull request by Dref360**
### Summary
Update the GeneratorEnqueuer logic to follow the Ordered Enqueuer (We could merge them later on)
* Remove the Lock and adds a new error message when use_multiprocessing=False.
* The exceptions raised are now similar between a Generator and a Sequence. (StopIteration)
* wait_time is now deprecated.

### Related Issues
https://github.com/keras-team/keras/pull/8662#discussion_r210682836

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [x] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

cc @datumbox @de-vri-es

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR. Goals and code look good to me.

We seem to be having test flakiness issues with TF currently. Possibly related to the recent TF 1.9 upgrade on CI.

**Comment by Dref360** [comment]
There are still issues with this PR on Python3, working on it.

**Comment by de-vri-es** [comment]
Some time ago I modified `GeneratorEnquer` to re-throw the original exception from a generator in the main thread/process (https://github.com/keras-team/keras/pull/8485). It looks like this PR reverts the behaviour to always throwing `StopIteration` in the main thread.

I'd really like if the behaviour from my PR can be preserved.

**Comment by datumbox** [comment]
First of all, your patch makes the Generator lightning fast. Good job!

Here are some benchmarks of GPU utilization before an after your patch (check IO and gpu-util on the last column). Removing the lock gives you a 5x speed improvement.

Before - Keras 2.2.2 (gpu starvation is observed):
https://ibb.co/mL7TyK

After - your branch:
https://ibb.co/i7rkCe

As for why the tests are stalling, I believe this has to do with your Queue. In our patch, we had to remove not only the lock but also the max-size of the queue. This is because a thread might get blocked while adding stuff in the queue (if it's full). When you call stop, these threads wait infinitely and that's why probably your tests are stalling. 

Here is a modified version of my test snippet I posted on the other PR that can help you see it:

```python
import time
from keras.preprocessing import image
from keras.utils.data_utils import GeneratorEnqueuer

it = image.ImageDataGenerator().flow_from_directory('/path/to/data', target_size=(224, 224), batch_size=512)

reader = GeneratorEnqueuer(it, use_multiprocessing=False)
reader.start(workers=16, max_queue_size=16)
g = reader.get()

start = time.time()
n = 100
for i in range(n):
        x=g.next()


total_time = time.time()-start
print('Total time %f sec; Average time: %f sec' % (total_time, total_time/n))


print('Stopping Queue')
reader.stop()

print('Finished')

```

Output:
```
Total time 31.848560 sec; Average time: 0.318486 sec
Stopping Queue^C
```

What I would do instead is use non-blocking Queue with not defined max-size. You can then use `qsize()` to get an approximate size of the queue and sleep if it exceeds the maximum. The queue can exceed the maximum defined value in this implementation but it's size can't get more than `MAX-QUEUE_SIZE+WORKERS-1`. I would also recommend adding some debug messages when the queue consistently gets empty and warn users for possible GPU starvation. This has been a huge pain for us in the past and due to the lack of messages, it's very hard to debug without modifying the keras code.

Try out the queue size and if that does not work let me know. I can debug the code for you on Monday as now I don't have the necessary tools to do it.

**Comment by datumbox** [comment]
Sorry I need to provide a clarification on my previous comment. When I talked about stalling tests, I was referring to the travis output for the [penultimate commit](https://travis-ci.org/keras-team/keras/jobs/417473415):
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
The tests of your subsequent commit did not stall but I think this was just a happy coincidence of threads not falling into blocking state. I believe that in the penultimate commit the execution of the `test_generator_enqueuer_threads` test got stack on `enqueuer.stop()` because it's the last test running on `gw0`. I think my comment on the queue size is still valid and the above snippet can help you reproduce the problem.

As for the reason why the last commit fails on `test_stateful_metrics` with the tolerance error, I have no clue. Might be related to TF as @fchollet said. On my machine the specific tests pass on your branch.

**Comment by Dref360** [comment]
@de-vri-es Now the behavior is similar to the Sequence behaviour, I wrap the exception in a StopIteration, we could rethrow as before. In any case, we should try to have a uniform behaviour.

@datumbox  in `stop`, we clear the queue so that the enqueuing threads get the final `put`. The thread will then die and kill the `mp.Pool`. I have a strong feeling that it's just the usual OOM from tensorflow since 1.9. I'll investigate the stateful metric test.
 When we call `stop`, the enqueuer shouldn't be used and the data shouldn't be trusted.

**Comment by datumbox** [comment]
@Dref360 I have not checked your modifications on an IDE so I just noticed that you clear the queue on `stop`. I'll check it more thoroughly next week. Does the example that I posted above hangs for you when`stop` is called? For me it did and I saw similar behaviour on [earlier commits](https://travis-ci.org/keras-team/keras/jobs/417473415) of your PR on Travis. I also see that the last commit passed the tests once your rerun it on travis. Sounds good on investigating the OOM, let us know your findings. If you don't spot anything there, we should look on the order of the data that I describe above.

Don't get me wrong, I like your changes; the Enqueuer was heavily patched by multiple PRs (myself included) and it was long due for someone to rationalize it. Just checking because multi-threading problems are hard to debug/reproduce.

**Comment by Dref360** [comment]
Your example works on my branch. I appreciate your help :P

**Comment by Dref360** [comment]
test_stateful_metrics is just flaky. Maybe having a "deterministic" model is something we should look for.

Nonetheless, I fixed a potential bug for the "generator" part. Even with `workers=1` the order is never guaranteed. Setting it to workers=0 guarantees the order.

**Comment by datumbox** [comment]
@Dref360 you are right. I have another unrelated PR and it randomly failed on `test_stateful_metrics`. This is not related.

**Comment by de-vri-es** [comment]
> @de-vri-es Now the behavior is similar to the Sequence behaviour, I wrap the exception in a StopIteration, we could rethrow as before. In any case, we should try to have a uniform behaviour.

Makes sense. I do think it's better to throw the original exception rather than a `StopIteration`. It significantly increased the quality of bug reports we got on our custom generators. Think that is possible here, or do you think a separate PR will be better?

**Comment by Dref360** [comment]
I agree with you, I'll make the changes in this PR this week.

**Comment by datumbox** [comment]
LGTM, good work.

**Comment by fchollet** [comment]
Tests looking stuck; restarting.

**Comment by fchollet** [comment]
Please resolve remaining merge conflicts.
